Mathlib/Tactic/HashCommandLinter.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Std.Lean.Command
+import Std.Data.Array.Basic
+
+/-!
+#  `#`-command linter
+
+The `#`-command linter produces a warning when a command starting with `#` is used.
+-/
+
+namespace Mathlib.Linter
+
+/--
+Setting to `true` enables the '`#`-command' linter.  When `true`, Lean warns on any command
+beginning with `#`.  For example, `#eval, #guard, #guard_msgs` all trigger a message.
+-/
+register_option linter.hashCommand : Bool := {
+  defValue := false
+  descr := "enable the `#`-command linter"
+}
+
+namespace HashCommandLinter
+
+open Lean Elab
+
+/-- `getAtomStx stx` extracts the array of all `.atom` nodes in `stx`. -/
+def getAtomStx : Syntax → Array Syntax
+  | .node _ _ args => (args.attach.map fun ⟨a, _h⟩ => getAtomStx a).foldl (· ++ ·) default
+  | s@(.atom ..) => #[s]
+  | _ => default
```

Comment:
This is a remnant of an earlier version of the linter, where I would list all atoms and scan for the ones beginning with `#` to try and do the "burrowing" that you mention above.  This is also why I later excluded spaces, since I wanted to discard matching on `# Section` in doc-modules.  I also removed the check for spaces below.